vms/platformvm/txs/executor/proposal_tx_preference.go

```diff
@@ -0,0 +1,179 @@
+)
+
+type ProposalTxPreference struct {
+	// inputs, to be filled before visitor methods are called
+	PrimaryUptimePercentage float64
+	Uptimes                 uptime.Calculator
+	State                   state.Chain
+	Tx                      *txs.Tx
+
+	// outputs populated by this struct's methods:
+	//
+	// [PrefersCommit] is true iff this node initially prefers to commit the
+	// block containing this transaction.
+	PrefersCommit bool
+}
+
+func (*ProposalTxPreference) CreateChainTx(*txs.CreateChainTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) CreateSubnetTx(*txs.CreateSubnetTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) ImportTx(*txs.ImportTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) ExportTx(*txs.ExportTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) RemoveSubnetValidatorTx(*txs.RemoveSubnetValidatorTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) TransformSubnetTx(*txs.TransformSubnetTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) AddPermissionlessValidatorTx(*txs.AddPermissionlessValidatorTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) AddPermissionlessDelegatorTx(*txs.AddPermissionlessDelegatorTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) TransferSubnetOwnershipTx(*txs.TransferSubnetOwnershipTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) BaseTx(*txs.BaseTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) AddValidatorTx(*txs.AddValidatorTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) AddSubnetValidatorTx(*txs.AddSubnetValidatorTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) AddDelegatorTx(*txs.AddDelegatorTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) AdvanceTimeTx(*txs.AdvanceTimeTx) error {
+	return ErrWrongTxType
+}
+
+func (e *ProposalTxPreference) RewardValidatorTx(tx *txs.RewardValidatorTx) error {
+	stakerTx, _, err := e.State.GetTx(tx.TxID)
+	if err == database.ErrNotFound {
+		// This can happen if this transaction is attempting to reward a
+		// validator that hasn't been confirmed.
+		e.PrefersCommit = true
+		return nil
+	}
+	if err != nil {
+		// GetTx can only return [ErrNotFound], or an unexpected error like a
+		// parsing error or internal DB error. For anything other than
+		// [ErrNotFound] the block can just be dropped.
+		return fmt.Errorf("%w %s: %w",
+			ErrMissingStakerTx,
+			tx.TxID,
+			err,
+		)
+	}
+
+	staker, ok := stakerTx.Unsigned.(txs.Staker)
+	if !ok {
+		// Because this transaction isn't guaranteed to have been verified yet,
+		// it's possible that a malicious node issued this transaction into a
+		// block that will fail verification in the future.
+		return fmt.Errorf("%w %s: %T",
+			ErrUnexpectedStakerTransactionType,
+			tx.TxID,
+			stakerTx.Unsigned,
+		)
```

Comment:
This can cause the node to FATAL if a malicious block producer includes a `TxID` for a non `txs.Staker` tx.